File: crates/idakit-sys-codegen/src/domains/cfg.rs
use super::super::model::*;

/// The control-flow-graph domain: the SDK's `qflow_chart_t` bound as an `Opaque` `ExternType`
/// (`FlowChart`) owned by [`UniquePtr`](cxx::UniquePtr), so its C++ deleter handles cleanup without
/// a manual free function or a hand-written `Drop` impl. `size` is a `self:`-member call bound straight to
/// `qflow_chart_t::size()` (no facade body); every other accessor is a free function over a
/// `&FlowChart`, hand-written in `facade/cfg.cpp`. Block bounds return by value as a `BlockInfo`
/// shared struct, and the successor/predecessor edge lists copy into owned `Vec<u32>`.
pub const CFG: Domain = Domain {
    name: "cfg",
    sdk_includes: &["<funcs.hpp>", "<gdl.hpp>", "<stdexcept>"],
    externs: &[ExternTy {
        rust_name: "FlowChart",
        cxx_name: "qflow_chart_t",
        kind: ExternKind::Opaque,
        doc: "The SDK's `qflow_chart_t`, an opaque control-flow graph handled only behind \
              indirection (`&FlowChart` or `UniquePtr<FlowChart>`).",
        safety: "The type id names the real SDK class qflow_chart_t; Opaque is correct because \
                 qflow_chart_t has a virtual destructor (nontrivial), so it may only cross the \
                 bridge behind a reference or UniquePtr, never by value.",
    }],
    structs: &[SharedStruct {
        name: "BlockInfo",
        doc: "One basic block's bounds and kind, returned by value from [`cfg_block`].",
        fields: fields! {
            start: U64 = "Start address of the block.";
            end: U64 = "End address (exclusive) of the block.";
            kind: I32 = "Raw `fc_block_type_t` discriminant (`fcb_normal`, `fcb_ret`, ...).";
        },
    }],
    consts: &[],
    custom_tus: &["facade/cfg.cpp"],
    fns: fns! {
        "Build the flow chart for the function containing `ea`; `Err` when no function is there. \
         Runs analysis, so it can also fail from a thrown SDK exception."
            cfg_build(ea: U64, flags: I32) -> ResultUniquePtr("FlowChart");
        "Number of basic blocks, bound to `qflow_chart_t::size()` directly (the `self:` receiver). \
         The return is `i32` to match the member's exact `int` signature."
            size(self: FlowChart) -> I32;
        "Total number of basic blocks (external blocks included)."
            cfg_nblocks(fc: ExternRef("FlowChart")) -> Usize;
        "Number of blocks belonging to the function's own range."
            cfg_nproper(fc: ExternRef("FlowChart")) -> Usize;
        "Bounds and kind of block `n`; `Err` when `n` is out of range."
            cfg_block(fc: ExternRef("FlowChart"), n: Usize) -> ResultShared("BlockInfo");
        "Number of successors of block `n` (`0` when `n` is out of range)."
            cfg_nsucc(fc: ExternRef("FlowChart"), n: Usize) -> Usize;
        "The `i`-th successor block index of block `n`; `Err` when `n`/`i` is out of range."
            cfg_succ(fc: ExternRef("FlowChart"), n: Usize, i: Usize) -> ResultUsize;
        "Number of predecessors of block `n` (`0` when `n` is out of range)."
            cfg_npred(fc: ExternRef("FlowChart"), n: Usize) -> Usize;
        "The `i`-th predecessor block index of block `n`; `Err` when `n`/`i` is out of range."
            cfg_pred(fc: ExternRef("FlowChart"), n: Usize, i: Usize) -> ResultUsize;
        "The whole successor edge list of block `n` as one owned `Vec<u32>`; `Err` when `n` is out \
         of range."
            cfg_succs(fc: ExternRef("FlowChart"), n: Usize) -> ResultVecU32;
        "The whole predecessor edge list of block `n` as one owned `Vec<u32>`; `Err` when `n` is \
         out of range."
            cfg_preds(fc: ExternRef("FlowChart"), n: Usize) -> ResultVecU32;
    },
};

File: crates/idakit-sys/facade/cfg.cpp
// Hand-written Custom bodies for the generated control-flow-graph domain (namespace gen).
// FlowChart is the SDK's qflow_chart_t owned by std::unique_ptr, so cxx's UniquePtr deleter retires
// the raw free + Rust Drop; cfg_block returns the BlockInfo shared struct by value, and the edge
// lists copy into rust::Vec<uint32_t>. size() is a self:-member call bound straight to the SDK
// member, so it has no body here. Out-of-range indices throw and surface as a Rust Err.

#include <ida.hpp>
#include <pro.h>

#include <funcs.hpp>
#include <gdl.hpp>

#include <stdexcept>

#include "gen_cfg.h"
// The generated bridge header defines the BlockInfo shared struct (full definition needed to
// construct it below); gen_cfg.h only forward-declares it.
#include "gen_bridge.h"

namespace gen {

std::unique_ptr<::qflow_chart_t> cfg_build(uint64_t ea, int32_t flags) {
  func_t *pfn = get_func((ea_t)ea);
  if (pfn == nullptr)
    throw std::out_of_range("no function at address");
  return std::make_unique<::qflow_chart_t>("", pfn, BADADDR, BADADDR, flags);
}

size_t cfg_nblocks(const ::qflow_chart_t &fc) { return fc.blocks.size(); }

size_t cfg_nproper(const ::qflow_chart_t &fc) { return (size_t)fc.nproper; }

BlockInfo cfg_block(const ::qflow_chart_t &fc, size_t n) {
  if (n >= fc.blocks.size())
    throw std::out_of_range("block index out of range");
  const qbasic_block_t &b = fc.blocks[n];
  BlockInfo info;
  info.start = (uint64_t)b.start_ea;
  info.end = (uint64_t)b.end_ea;
  info.kind = (int32_t)fc.calc_block_type(n);
  return info;
}

size_t cfg_nsucc(const ::qflow_chart_t &fc, size_t n) {
  if (n >= fc.blocks.size())
    return 0;
  return (size_t)fc.nsucc((int)n);
}

size_t cfg_succ(const ::qflow_chart_t &fc, size_t n, size_t i) {
  if (n >= fc.blocks.size() || i >= (size_t)fc.nsucc((int)n))
    throw std::out_of_range("successor index out of range");
  return (size_t)fc.succ((int)n, (int)i);
}

size_t cfg_npred(const ::qflow_chart_t &fc, size_t n) {
  if (n >= fc.blocks.size())
    return 0;
  return (size_t)fc.npred((int)n);
}

size_t cfg_pred(const ::qflow_chart_t &fc, size_t n, size_t i) {
  if (n >= fc.blocks.size() || i >= (size_t)fc.npred((int)n))
    throw std::out_of_range("predecessor index out of range");
  return (size_t)fc.pred((int)n, (int)i);
}

// Copy the block's succ/pred intvec_t (qvector<int>) into an owned rust::Vec<uint32_t>: one linear
// copy, no lifetime tie to the FlowChart.
rust::Vec<uint32_t> cfg_succs(const ::qflow_chart_t &fc, size_t n) {
  if (n >= fc.blocks.size())
    throw std::out_of_range("block index out of range");
  const intvec_t &succ = fc.blocks[n].succ;
  rust::Vec<uint32_t> out;
  out.reserve(succ.size());
  for (int s : succ)
    out.push_back((uint32_t)s);
  return out;
}

rust::Vec<uint32_t> cfg_preds(const ::qflow_chart_t &fc, size_t n) {
  if (n >= fc.blocks.size())
    throw std::out_of_range("block index out of range");
  const intvec_t &pred = fc.blocks[n].pred;
  rust::Vec<uint32_t> out;
  out.reserve(pred.size());
  for (int p : pred)
    out.push_back((uint32_t)p);
  return out;
}

} // namespace gen
